### ISSUE TITLE
Forward mode autodiff support

### ISSUE THREAD

**Issue by crowsonkb** [SHI-Labs/NATTEN#68]
I have a use case for forward mode autodiff (JVP) support, and support for backprop through the JVP, for NATTEN. It is for distilling diffusion models that use NATTEN (paper in progress) into [consistency models](https://arxiv.org/abs/2303.01469).

I think the JVP of the NATTEN ops can be expressed in terms of the existing CUDA kernels so there should be no need to write new kernels for this. The tangent of `natten2dqk(q_primal, k_primal)`, for instance, is `natten2dqk(q_tangent, k_primal) + natten2dqk(q_primal, k_tangent)`, the same pattern holds for natten2dav and the kernels for other numbers of dimensions. (I do not know what happens with the rpb but I don't use it, I use RoPE instead.) Since you provide backward passes for these this also provides the ability to backprop through the JVP output, which I need.

Thank you,
Katherine Crowson

**Comment by alihassanijr** [comment]
Thank you for your interest and for sharing this.

I'd just add one note in case you find it useful; if we exclude RPB, there's only 3 different operations that implement everything. One op can compute the QK operation, as well as the gradient for attention weights, one can compute AV (attention weights applied to values) and the gradient for Q, and the third op computes the gradients for K and V. The only difference between these three is just their inputs.

While their underlying kernels are the same, we don't directly expose the ops through the API, and instead expose them as needed for a full forward pass and backward pass.

If you think exposing these directly, or making other such changes (i.e. stacking the two terms in order to compute them with one kernel call and then reducing) is going to be helpful, please let us know.

I've been away from the project for a few months, but I'm coming back to it in a few weeks, making improvements to both the API, torch package, and of course kernels.

**Comment by confucianism72** [comment]
> I do not know what happens with the rpb but I don't use it, I use RoPE instead.

In my opinion, rpb here is also one learnable positional embedding. It is added on attention matrix.
Just like 

$$
q_i^\top k_j= x_i^\top W_q W_k x_j + b_{ij}
$$

**Comment by confucianism72** [comment]
size of bias is small in NT, roughly the same size as kernel. the implementation is efficient I think

**Comment by alihassanijr** [comment]
Yes, RPB is roughly twice the size of the kernel, in order to capture all possible offsets given the kernel size (because the query is not always centered, otherwise it would be exactly the size of the kernel).

**Comment by Birch-san** [comment]
@alihassanijr  
> If you think exposing these directly, or making other such changes (i.e. stacking the two terms in order to compute them with one kernel call and then reducing) is going to be helpful, please let us know.

the main ask is just to register the `natten2dqk` and `natten2dqk` operations as torch autograd operations which declare a `jvp()` static method:  
https://pytorch.org/tutorials/intermediate/forward_ad_usage.html#custom-autograd-function

the library already exposes enough that a user can already declare this themselves, but it's better for it to live inside the library.

of course, stacking the two terms into one kernel call would be more performant, so if you can see a way to do that then that's helpful too. but exposing a naïve implementation would be a good starting point.

the nice thing about supporting forward-mode autodiff, is that a vision model can output channels from which the second derivative of the diffusion ODE can be computed. you could either use this for more efficient diffusion sampling (using the second derivative for a better estimate of the trajectory of the ODE), or you could use this for consistency training (encouraging the second derivative to be 1, so that you can follow the trajectory to the end without requiring many changes in direction — i.e. few-step sampling)

**Comment by alihassanijr** [comment]
I see; thanks for bringing it up. I originally didn't realize torch had apis for JVP.
I'll definitely want to add support for this to NATTEN.

I'll look into this; but in the meantime any reference implementation would be appreciated.
The single kernel call might be a good idea for some cases; but if a tangent and primal pair is not already in contiguous memory, then we might have to concat the two, which might not end up being that much better than just calling the ops twice. But we can cross this bridge later on.

**Comment by Birch-san** [comment]
@alihassanijr I implemented jvp for `natten2dqk` and `natten2dav` according to @crowsonkb's derivation, and its result satisfies allclose() with pure-pytorch neighbourhood attention (torch sdp + Kat's attention mask):  
https://github.com/Birch-san/natten-fwd-ad/blob/main/src/natten_autograd.py

Test case:  
https://github.com/Birch-san/natten-fwd-ad/blob/main/script/demo.py

this "test against pure-pytorch" approach could also be useful for verifying a nested tensor implementation. 🙂

**Comment by alihassanijr** [comment]
Thank you @Birch-san ! Looks like the JVP-compatible functions will have a different signature in their respective autograd functions, so I'll probably have to separate the two out, but hopefully I should be done soon.

And I haven't been able to test the nested API yet, even though most of the refactoring in the C++ API that was necessary is done. I'll see if I can get both of these out by the end of this week. I appreciate your help and patience.

**Comment by Birch-san** [comment]
looks like merging the PR didn't close the issue.

should we close it now (and if we find bugs after release, report them then)?

**Comment by alihassanijr** [comment]
Feel free to keep it open until you get a chance to test; either way is fine.

**Comment by Birch-san** [comment]
ah yeah, I did indeed get a chance to test this (same time as I was testing nested tensors):  
https://github.com/SHI-Labs/NATTEN/pull/76#issuecomment-1872603778

your forward-mode kernel worked for me when I compiled NATTEN locally and just used the usual ops :)

**Comment by Birch-san** [comment]
yup, should be fine to close this now. 🙂 
looking forward to the release! thanks for your work.
